---
title: "R Notebook"
output: html_notebook
---

# 1. Downloading and Preprocessing Data


```{r, results = FALSE}
### Potentially, we will need the developer version of some packages:

#devtools::install_github("BioinformaticsFMRP/TCGAbiolinks")
#BiocManager::install("BioinformaticsFMRP/TCGAbiolinksGUI.data")
#BiocManager::install("BioinformaticsFMRP/TCGAbiolinks")

### Set Seed and load packages
set.seed(123)
library("TCGAbiolinks")
library("tidyr")
library("limma")
library("edgeR")
library("dplyr")
library("ggplot2")
library("DESeq2")
library("flipscores")
library("MASS")

### Load Helper Functions
source(paste0("../functions/", "printProgress.R"))

### Specify which parts of the analysis are reproduced
reproAll <- FALSE
reproList <- list("downloadingRawData" = FALSE | reproAll,
                   "computingTCGALIHCResults" = FALSE | reproAll)

### Specify Directories
dataDir <- "../data/"

rm(reproAll)
```
## 1.1 Downloading and Saving Data

```{r}
### First check if data already exists locally.
### If we do not want to reproduce the download step,  load it from there
if (file.exists(paste0(dataDir, "TCGALIHC.Rds")) | reproList$downloadingRawData){
  message("Load Saved Data")
  TCGAData <- readRDS(file = paste0(dataDir, "TCGALIHC.RDS"))
  

} else {
  message("Download and Save Data")
  ### Search Data in TCGA Data Base
  queryTCGA = TCGAbiolinks::GDCquery(
    project = "TCGA-LIHC",
    data.category = "Transcriptome Profiling", 
    data.type = "Gene Expression Quantification", 
    workflow.type = "STAR - Counts"
)
  ### Download TCGA Data (saved locally)
  TCGAbiolinks::GDCdownload(query = queryTCGA, 
                            directory = dataDir,
                            files.per.chunk = 50)
  
  ### Read and Prepare Downloaded Data
  TCGAData <- TCGAbiolinks::GDCprepare(queryTCGA,
                                          directory = dataDir)

  TCGAData <- list(counts = assay(TCGAData),
                      patient = TCGAData$patient,
                      group = TCGAData$definition,
                      sampleType = TCGAData$sample_type)
  ### Filter relevant data and Save Data to Specified path
  saveRDS(object = TCGAData,
          file = paste0(dataDir, "TCGALIHC.Rds"))
  
  ### Delete Unprepared Data
  unlink(paste0(dataDir, "TCGA-LIHC"), recursive=TRUE)
  rm(queryTCGA)
}
```



```{r Check data}
### Create Mask with duplicate Samples
duplicatedMask <- duplicated(TCGAData$patient)
sum(duplicatedMask)
patientMask <- TCGAData$patient %in% TCGAData$patient[duplicatedMask]

### The duplicate samples are coming from paired data, for which we delete
### the samples corresponding to the tumor.
patientMaskDelete <- patientMask & TCGAData$group == "Primary solid Tumor"

### Let us now compute the normalization factors and plot the MDS of the data

#### Compute EdgeR object
dge   <- DGEList(counts = TCGAData$counts[,!patientMaskDelete],
                 group  = TCGAData$group[!patientMaskDelete])
keep  <- filterByExpr(dge)
dge   <- dge[keep,,keep.lib.sizes=FALSE]
dge   <- calcNormFactors(dge)

#### Compute plot and save it
fileName <- paste("../plots/", "mdsTCGALIHC", ".png", sep = "")
MDSData <- limma::plotMDS(dge, labels = 1:ncol(dge), col = as.numeric(dge$samples$group), plot = F)
(MDSPlot <- ggplot(data = as.data.frame(list(x = MDSData$x, y = MDSData$y,
                                       group = as.factor(dge$samples$group))), 
             mapping = aes(x = x, y = y, col = group)) + 
  geom_point() +
  labs(x = paste0("Leading logFC dim 1 (", round(MDSData$var.explained[1],3)*100, "%)"),
       y = paste0("Leading logFC dim 2 (", round(MDSData$var.explained[2], 3)*100, "%)"), 
       title = "MDS of TCGA-LIHC Data")
  )

ggsave(fileName, MDSPlot)


### Remove Third Sample Type (Recurrent Tumor) and save processed data
TCGAData$patientMaskDelete <- patientMaskDelete | TCGAData$sampleType == "Recurrent Tumor"


saveRDS(object = TCGAData,
          file = paste0(dataDir, "TCGALIHC.Rds"))

rm(dge, MDSData, MDSPlot, TCGAData, duplicatedMask, keep, patientMask, 
   patientMaskDelete)
```



# Ch.4: Parametric Methods
Now, let us run all parametric models. First, important the packages.
```{r Setup Parametric Methods, echo = F}
### Load Data and filter
tcgaDataCleaned <-tcgaData <- readRDS(paste0(dataDir, "TCGALIHC.Rds"))
tcgaDataCleaned$counts <- tcgaData$counts[,!tcgaData$patientMaskDelete]
tcgaDataCleaned$group  <- tcgaData$group[!tcgaData$patientMaskDelete]
```
 


```{r Limma}
### Initialize DGEList Object
design <-  model.matrix(~ tcgaDataCleaned$group)
dge    <-  edgeR::DGEList(counts=tcgaDataCleaned$counts)

### Normalization (TMM followed by voom w. different spans)
dge   <- edgeR::calcNormFactors(dge)
v01   <- limma::voom(dge, design, plot=F, save.plot = T, span = 0.1)
v02   <- limma::voom(dge, design, plot=F, save.plot = T, span = 0.2)
v05   <- limma::voom(dge, design, plot=F, save.plot = T, span = 0.5)

### Filter genes, normalize again and again compute voom w. different spans
keep          <- edgeR::filterByExpr(dge, design)
dgeFiltered   <- dge[keep,,keep.lib.sizes=FALSE]
dge           <- edgeR::calcNormFactors(dge)
v01Filtered   <- limma::voom(dgeFiltered, design, plot=F, save.plot = T, span = 0.1)
v02Filtered   <- limma::voom(dgeFiltered, design, plot=F, save.plot = T, span = 0.2)
v05Filtered   <- limma::voom(dgeFiltered, design, plot=F, save.plot = T, span = 0.5)

### Compute p-values and pad filtered out genes with NAs
fit   <- limma::lmFit(v01Filtered, design)
fit   <- limma::eBayes(fit, robust = T)
pVals <- data.frame("pValue" = fit$p.value[,2])
df    <- data.frame(row.names = rownames(tcgaData$counts),
                    pValue = rep(NA, nrow(tcgaData$counts)))

df[rownames(df) %in% rownames(pVals), 1] <- pVals$pValue

### Save p-values and write data to disk
tcgaData$Results$Limma <- df$pValue
saveRDS(tcgaData, paste0(dataDir, "TCGALIHC.Rds"))

##### PLOT LIMMA Lines #####

### Bind all objects together
linesDf <- dplyr::bind_rows(v01$voom.line, v02$voom.line, v05$voom.line,
                             v01Filtered$voom.line,
                             v02Filtered$voom.line,
                             v05Filtered$voom.line)
### Append data for plotting
linesDf$span <- c(rep(as.character(c(0.1, 0.2, 0.5)), 
                       each = length(v01$voom.line$x)),
                   rep(as.character(c(0.1, 0.2, 0.5)), 
                       each = length(v02Filtered$voom.line$x))
                   )
linesDf$filter <- c(rep(F, 3*length(v01$voom.line$x)),
                     rep(T, 3*length(v02Filtered$voom.line$x))
                     )

### Plot Data
(linesPlot <- ggplot2::ggplot() +
  geom_point(mapping = aes(x = v01$voom.xy$x, y = v01$voom.xy$y)) +
  geom_line(mapping = aes(x = linesDf$x, y = linesDf$y, col = linesDf$span,
                          linetype = linesDf$filter)) +
  labs(x = v01$voom.xy$xlab,
       y = v01$voom.xy$ylab,
       title = "voom: Mean-Variance Trend for different Spans",
       col = "Span",
       linetype = "Filter")
)

### Save Plot
fileName <- paste("../plots/voomLine", ".png", sep = "")
ggsave(fileName, linesPlot, width = 16, height = 8, units = "cm")

rm(design, df, dge, dgeFiltered, fit, linesDf, linesPlot, pVals, v01,
   v01Filtered, v02, v02Filtered, v05, v05Filtered, keep, fileName)
```


```{r EdgeR Exact}
### Initialize EdgeR object
design <- model.matrix(~ tcgaDataCleaned$group)
dge    <- edgeR::DGEList(counts=tcgaDataCleaned$counts,
                         group = design[,2])

### Run EdgeR Pipeline and pad filtered out genes with NA
dge <-  edgeR::calcNormFactors(dge)
dgeDefault <-  edgeR::estimateDisp(dge)
dgeUnshrunken <- edgeR::estimateDisp(dge, prior.df = 1e-10)
et  <-  edgeR::exactTest(dgeDefault) 


### Save p-values and write data to disk
tcgaData$Results$EdgeREt <- et$table$PValue
saveRDS(tcgaData, paste0(dataDir, "TCGALIHC.Rds"))

### Plot Overdispersion vs logCPM
(BCVPlot <- ggplot() +
  geom_point(alpha = 0.35, aes(x = dgeDefault$AveLogCPM, y = sqrt(dgeDefault$tagwise.dispersion))) +
  geom_line(aes(x = dgeDefault$AveLogCPM, y = sqrt(dgeDefault$trended.dispersion), col = "Trend")) +
  geom_hline(aes(yintercept = sqrt(dgeDefault$common.dispersion), col = "Common")) +
  labs(x = "Average log CPM",
       y = "BCV",
       title = "Overdispersion vs. Log-CPM")
)

### Save Plot
fileName <- paste("../plots/EdgeRETLine", ".png", sep = "")
ggsave(fileName, BCVPlot, width = 16, height = 8, units = "cm")

rm(design, dge, et, BCVPlot, fileName)
```


```{r EdgeR histogram}
### Generate Data
y <- matrix(rnbinom(50000*4, mu=100, size=1/0.25), ncol=4)
group <- factor(c(1,1,2,2))
design <- model.matrix(~group)
d <- DGEList(counts=y, group=group)

### Run Test by plugging in true dispersion
pTrue <- exactTest(d, dispersion = 0.25, rejection.region = "doubletail")$table$PValue
plotTrue  <- ggplot(mapping = aes(x = pTrue)) + 
  geom_histogram(bins = 20) +
  labs(x = "p")

### Run Test using estimated dispersions
d1 <- estimateDisp(d, prior.df = 0, trend.method = "none")
pEst  <- exactTest(d1, rejection.region = "doubletail")$table$PValue

### Plot Data
plotEst  <- ggplot(mapping = aes(x = pEst)) + 
  geom_histogram(bins = 20) +
  labs(x = "p")
plotBins <- ggplot(mapping = aes(x = pEst, y = sqrt(d1$tagwise.dispersion))) +
  geom_bin2d(bins = 16) +
  geom_hline(yintercept = 0.25, col = "red") +
  scale_y_sqrt() +
  labs(y = "BCV", x = "p") +
  scale_fill_gradient(trans = "sqrt",
                      breaks = 2^c(3, 7, 9, 11), 
                      labels = 2^c(3, 7, 9, 11))

p4 <- ggplot(mapping = aes(x = sqrt(dgeUnshrunken$tagwise.dispersion), y = sqrt(dgeDefault$tagwise.dispersion))) +
  geom_point(alpha  = 0.15) +
  geom_abline(a = 0, b = 1, col = "red") +
  labs(x = "Individual BCV", y = "Shrunken BCV")

(pAll <- cowplot::plot_grid(plotTrue, plotBins, plotEst, p4, nrow = 2,
                               align = "v", axis = "r", labels = "auto")
)
### Save Plot
filename <- paste("../plots/", "EdgeRPdist" , ".png", sep = "")
cowplot::save_plot(pAll, filename = filename, nrow =2, base_width = 7, base_height = 2)

rm(d, d1, design, dgeUnshrunken, p4, pAll, plotBins, plotEst, plotTrue, y, group, pEst, pTrue)
```


```{r EdgeR QLM Test}
design <- model.matrix(~ tcgaDataCleaned$group)
### Normalize Data, Fit Model and Compute tests
fit <- glmQLFit(dgeDefault, design)
qlf <- glmQLFTest(fit, coef=2)

### Save Data
tcgaData$Results$EdgeRQl <- qlf$table$PValue
saveRDS(tcgaData, paste0(dataDir, "TCGALIHC.Rds"))
```

```{r DESeq2}
### Initialize Object
Y   <- data.frame("treatment" = as.factor(tcgaDataCleaned$group))
dds <- DESeq2::DESeqDataSetFromMatrix(tcgaDataCleaned$counts,
                                      colData = Y,
                                      design = ~ treatment)
### Run DESeq2 Pipeline and Extract Results
dds <- DESeq2::DESeq(dds, quiet = F)
topGenes <- as.data.frame(DESeq2::results(dds, alpha = 1-1e-15,
                                          pAdjustMethod = "none",
                                          independentFiltering=FALSE)
                          )

tcgaData$Results$DESeq2 <- topGenes$pvalue
saveRDS(tcgaData, paste0(dataDir, "TCGALIHC.Rds"))
```

# Ch. 5 Non-Parametric Methods

```{r MW Test}
### Normalize Data
groupMask <-  tcgaDataCleaned$group == tcgaDataCleaned$group[1] #Makes Data Binary
dge <- edgeR::DGEList(counts= tcgaDataCleaned$counts)
dge <- edgeR::calcNormFactors(dge, method="TMM")

### compute normalized CPM, rounded to account for noise in ties
dgeCpm <- round(edgeR::cpm(dge), 1)

### Apply MWU Test to All Genes using the created mask and Save Data
tcgaData$Results$MWU <- sapply(1:nrow(dgeCpm), function(i) {
    wilcox.test(dgeCpm[i, groupMask], dgeCpm[i, !groupMask])$p.value
})
saveRDS(tcgaData, paste0(dataDir, "TCGALIHC.Rds"))
```


```{r Direct Permutation Test}
nFlips <- 2000
rn <- row.names(tcgaDataCleaned$counts)

### Normalize Data
d  <- edgeR::DGEList(counts = tcgaDataCleaned$counts)
d  <- edgeR::calcNormFactors(d)
os <- edgeR::getOffset(d)
X  <- d$counts * d$samples$norm.factors
Y  <- tcgaDataCleaned$group == tcgaDataCleaned$group[1]

t0 <- Sys.time()
pVals <- sapply(1:nrow(tcgaDataCleaned$counts), function(i){
    ### print progress w. time estimate
    if(i %% 50 == 0){PrintProgress(i, nrow(X), t0)}

    meanDiffs    <- numeric(nFlips)
    meanDiffs[1] <- abs(mean(X[i,Y == T]) - mean(X[i,Y == F]))
    for (j in 2:nFlips){
      yj <- sample(Y)
      meanDiffs[j] <- abs(mean(X[i,yj == T]) - mean(X[i,yj  == F]))
    }
    pPerm <- (sum(meanDiffs[1] <= meanDiffs))/nFlips
    return(pPerm)
  })

tcgaData$Results$DP <- pVals
saveRDS(tcgaData, paste0(dataDir, "TCGALIHC.Rds"))

rm(d, dds, dge, dgeDefault, X, pVals)
```


```{r Flipscores}
### Compute Offsets using EdgeR
rn <- row.names(tcgaDataCleaned$counts)
d  <- edgeR::DGEList(counts = tcgaDataCleaned$counts)
d  <- edgeR::calcNormFactors(d)
os <- edgeR::getOffset(d)

### Prepare Data
Y <- (tcgaDataCleaned$group == tcgaDataCleaned$group[1])*2-1 # Data being encoded (-1, 1)


### Loop through Columns and apply Flipscores test to each
t0 <- Sys.time()
pVals <- sapply(1:nrow(tcgaDataCleaned$counts), 
                 function(i) {
                   ### print progress
                   if(i %% 50 == 0){PrintProgress(i, nrow(tcgaDataCleaned$counts), t0)}
                   X_i <- d$counts[i,] 
                   suppressWarnings({
                   mod0 <- flipscores::flipscores(X_i ~ Y + offset(os),
                                                  family = poisson,
                                                  score_type = "standardized",
                                                  n_flips = 400,                #ca. 6h
                                                  to_be_tested = 2)
                   })
                   coef(summary(mod0))[2, 5] #  extract p-value from summary
                   }
                 )
print(Sys.time() - t0)

tcgaData$Results$FsPS <- pVals
saveRDS(tcgaData, paste0(dataDir, "TCGALIHC.Rds"))

rm(d, os, pVals, rn, t0, Y)
```

# Visualizing Results
```{r Setup Visualization}
namesList <- names(tcgaData$Results)
LIHCRes <- as.data.frame(tcgaData$Results)

colnames(LIHCRes) <- namesList

### Compute Rejected Genes and print number of Rejections for all Methods
LIHCResDE <- LIHCRes <= 0.05
LIHCResDE[is.na(LIHCResDE)] <- F
colSums(LIHCResDE, na.rm = T)
ftable(as.data.frame(LIHCResDE))
```


```{r}
corSpear <- cor(LIHCRes, method = "spearman", use  = "pairwise.complete.obs")
xtable::xtable(as.matrix(corSpear)) ##return latex table
```
```{r}
as.matrix(corSpear)
```


```{r}
mds <- MASS::isoMDS(1- corSpear +0.0001)
ggplot(mapping = aes(x = mds$points[,1], y = mds$points[,2], col = row.names(mds$points))) +
  geom_point()
```


